// @dcloudio/vite-plugin-uni/dist/config/css.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createCss = void 0;
const path_1 = __importDefault(require("path"));
const fs_extra_1 = __importDefault(require("fs-extra"));
function resolveAdditionalData(inputDir) {
    const uniScssFile = path_1.default.resolve(inputDir, 'uni.scss');
    if (!fs_extra_1.default.existsSync(uniScssFile)) {
        return '';
    }
    return fs_extra_1.default.readFileSync(uniScssFile, 'utf8');
}
function resolvePostcssConfig(inputDir) {
    return [
        path_1.default.resolve(inputDir, 'postcss.config.js'),
        path_1.default.resolve(process.cwd(), 'postcss.config.js'),
    ].find((file) => fs_extra_1.default.existsSync(file));
}
function createCss(options) {
    return {
        postcss: resolvePostcssConfig(options.inputDir),
        preprocessorOptions: {
            scss: {
                charset: false,
                additionalData: resolveAdditionalData(options.inputDir),
            },
        },
    };
}
exports.createCss = createCss;
